Apps/Common/src/Delegates/HMACHashDelegate.cs

```diff
@@ -0,0 +1,148 @@
+            this.HashConfig = new HMACHashDelegateConfig();
+            this.configuration.Bind(ConfigKey, this.HashConfig);
+        }
+
+        /// <summary>
+        /// Gets or sets the instance configuration.
+        /// </summary>
+        public HMACHashDelegateConfig HashConfig { get; set; }
+
+        /// <summary>
+        /// Generates the HMAC Hash object.
+        /// </summary>
+        /// <param name="key">The string to hash.</param>
+        /// <param name="salt">The salt to use for hashing.</param>
+        /// <param name="prf">The pseudorandom function to use for the hash.</param>
+        /// <param name="iterations">The number of iterations to process over the hash.</param>
+        /// <returns>The hash object.</returns>
+        public static HMACHash HMACHash(
+            string key,
+            byte[] salt,
+            KeyDerivationPrf prf = HMACHashDelegateConfig.DefaultPseudoRandomFunction,
+            int iterations = HMACHashDelegateConfig.DefaultIterations)
+        {
+            HMACHash retHash = new HMACHash()
+            {
+                PseudoRandomFunction = Database.Constant.HashFunction.HMACSHA512,
+                Iterations = iterations,
+            };
+
+            // Calculate the length in bytes of the hash given the function size
+            int hashLength = prf == KeyDerivationPrf.HMACSHA1 ? 20 : prf == KeyDerivationPrf.HMACSHA256 ? 32 : 64;
+            retHash.Salt = Convert.ToBase64String(salt);
+            retHash.Hash = Convert.ToBase64String(KeyDerivation.Pbkdf2(
+                                                    password: key,
+                                                    salt: salt,
+                                                    prf: prf,
+                                                    iterationCount: iterations,
+                                                    numBytesRequested: hashLength));
+            return retHash;
+        }
+
+        /// <summary>
+        /// Validates that the supplied key will hash to the compareHash.
+        /// </summary>
+        /// <param name="key">The key to hash and compare.</param>
+        /// <param name="compareHash">The hash object to compare.</param>
+        /// <returns>true if the key generates the same hash.</returns>
+        public static bool Compare(string key, HMACHash? compareHash)
+        {
+            bool result = false;
+            if (compareHash != null && compareHash.Hash != null && compareHash.Salt != null)
+            {
+                HMACHash keyHash = HMACHash(
+                    key,
+                    Convert.FromBase64String(compareHash.Salt),
+                    (KeyDerivationPrf)compareHash.PseudoRandomFunction,
+                    compareHash.Iterations);
+                result = compareHash.Hash == keyHash.Hash;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Generatea a secure random salt of the specified length in bytes.
```

Comment:
nit, Generatea -> Generate

Reply:
I'll fix this post merge for my next PR.